trieste/models/gpflux/sampler.py

```diff
@@ -191,3 +209,353 @@ def sample(self, at: TensorType, *, jitter: float = DEFAULTS.JITTER) -> TensorTy
+    underlying :class:`~gpflux.models.DeepGP` models by using a feature decomposition using both
+    random Fourier features and canonical features centered at inducing point locations. This allows
+    for cheap approximate trajectory samples, as opposed to exact trajectory sampling, which scales
+    cubically in the number of query points.
+
+    Note that we do not currently support deep GP models with multiple outputs.
+    """
+
+    def __init__(
+        self,
+        model: GPfluxPredictor,
+        num_features: int = 1000,
+    ):
+        """
+        :param model: The model to sample from.
+        :param num_features: The number of random Fourier features to use.
+        :raise ValueError (or InvalidArgumentError): If the model is not a :class:`GPfluxPredictor`,
+            or its underlying ``model_gpflux`` is not a :class:`~gpflux.models.DeepGP`, or
+            ``num_features`` is not positive.
+        """
+        if not isinstance(model, GPfluxPredictor):
+            raise ValueError(
+                f"Model must be a gpflux.interface.GPfluxPredictor, received {type(model)}"
+            )
+        if not isinstance(model.model_gpflux, DeepGP):
+            raise ValueError(
+                f"GPflux model must be a gpflux.models.DeepGP, received {type(model.model_gpflux)}"
+            )
+
+        super().__init__(model)
+        tf.debugging.assert_positive(num_features)
+        self._num_features = num_features
+        self._model_gpflux = model.model_gpflux
+        self._sampling_layers = [
+            DeepGaussianProcessDecoupledLayer(layer, num_features)
+            for layer in self._model_gpflux.f_layers
+        ]
+
+    def __repr__(self) -> str:
+        """"""
+        return f"""{self.__class__.__name__}(
+        {self._model!r},
+        {self._num_features!r})
+        """
+
+    def get_trajectory(self) -> TrajectoryFunction:
+        """
+        Generate an approximate function draw (trajectory) from the deep GP model.
+
+        :return: A trajectory function representing an approximate trajectory from the deep GP,
+            taking an input of shape `[N, B, D]` and returning shape `[N, B]`.
+        """
+
+        return dgp_feature_decomposition_trajectory(self._sampling_layers)
+
+    def update_trajectory(self, trajectory: TrajectoryFunction) -> TrajectoryFunction:
+        """
+        Efficiently update a :const:`TrajectoryFunction` to reflect an update in its underlying
+        :class:`ProbabilisticModel` and resample accordingly.
+
+        :param trajectory: The trajectory function to be updated and resampled.
+        :return: The updated and resampled trajectory function.
+        :raise InvalidArgumentError: If ``trajectory`` is not a
+            :class:`dgp_feature_decomposition_trajectory`
+        """
+
+        tf.debugging.Assert(isinstance(trajectory, dgp_feature_decomposition_trajectory), [])
+
+        cast(dgp_feature_decomposition_trajectory, trajectory).update()
```

Comment:
should there be a `resample()` call after the `update()` call?

Reply:
No, it's fine - this calls `update()` in `dgp_feature_decomposition_trajectory`, which then calls `update()` in each layer, which makes sure weights are resampled.